scarlet/lite/component.py

```diff
@@ -0,0 +1,621 @@
+        prox_sed: `Function`
+            Proximal operator for the SED.
+            If `prox_sed` is `None` then the default proximal
+            operator `self.prox_sed` is used.
+        floor: `float`
+            The minimum value of the SED, used to prevent
+            divergences in the gradients.
+        """
+        params = FixedParameter(morph_params)
+        sed = FixedParameter(sed)
+
+        self._params = params
+        self._func = morph_func
+        self._morph_grad = morph_grad
+        self._morph_prox = morph_prox
+        self._morph_step = morph_step
+        self._sed = sed
+        self._bbox = bbox
+        if prox_sed is None:
+            self._prox_sed = self.prox_sed
+        else:
+            self._prox_sed = prox_sed
+        self.slices = overlapped_slices(model_frame.bbox, bbox)
+        self.floor = floor
+
+    @property
+    def center(self):
+        """The center of the component"""
+        return self.y0, self.x0
+
+    @property
+    def y0(self):
+        """The y-center of the component"""
+        return self._params.x[0]
+
+    @property
+    def x0(self):
+        """The x-center of the component"""
+        return self._params.x[1]
+
+    @property
+    def sed(self):
+        """The SED of the component"""
+        return self._sed.x
+
+    @property
+    def bbox(self):
+        """The bounding box that contains the component"""
+        return self._bbox
+
+    @property
+    def frame(self):
+        """The coordinate system that contains the model"""
+        return CartesianFrame(self._bbox)
+
+    @property
+    def radial_params(self):
+        """The parameters used to model the radial function"""
+        return self._params.x
+
+    def _morph(self, frame=None):
+        if frame is None:
+            frame = self.frame
+        return self._func(self.radial_params, frame)
+
+    @property
+    def morph(self, frame=None):
+        """The morphological model"""
+        return self._morph()
+
+    @property
+    def morph_prox(self):
+        """The function used to constrain the morphological model"""
+        return self._morph_prox
+
+    @property
+    def morph_grad(self):
+        """The function that calculates the gradient of the morphological model"""
+        return self._morph_grad
+
+    @property
+    def morph_step(self):
+        """The function that calculates the gradient of the morphological model"""
+        return self._morph_step
+
+    def get_model(self, bbox=None, frame=None):
+        """Generate the full model for this component"""
+        model = self.sed[:, None, None] * self._morph(frame)[None, :, :]
+
+        if bbox is not None:
+            slices = overlapped_slices(bbox, self.bbox)
+            _model = np.zeros(bbox.shape, self.morph.dtype)
+            _model[slices[0]] = model[slices[1]]
+            model = _model
+        return model
+
+    def prox_sed(self, sed, prox_step=0):
+        """Apply a prox-like update to the SED
```

Comment:
Most of the proxes that we use in scarlet are not true proximal operators, due to the time it would take to generate them correctly in the true L2 sense. Instead we often use projection operators that are prox-like but not necessary the optimal operator to satisfy the constraint.